packages/lesswrong/server/migrations/2020-05-06-forceSecureImageLinks.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Turn insecure image links into secure ones if the secure image exists. We
+ * only edit the html (not the original content), so it's possible someone will
+ * edit their post and the update will be reverted.
+ */
+
+import { registerMigration, forEachDocumentBatchInCollection } from './migrationUtils';
+import { urlIsBroken } from '../scripts/utils'
+import { Posts } from '../../lib/collections/posts/collection';
+import cheerio from 'cheerio'
+
+function findInsecureImages ($: any): Array<string> {
+  let insecureImageSources: Array<string> = []
+  $('img').each(function (this: any) {
+    const el = $(this)
+    const src = el.attr('src') as string
+    const testResult = /^http:\/\//.test(src)
+    if (testResult) {
+      insecureImageSources.push(src)
+    }
+  })
+  return insecureImageSources
+}
+
+async function testSecureImages
+  (insecureImageSources: Array<string>, post: DbPost):
+  Promise<{shouldUpdate: boolean, imageUpdates: Map<string, string>}>
+{
+  let shouldUpdate = false
+  const imageUpdates = new Map((await Promise.all(
+    insecureImageSources.map(async (src): Promise<[string, string | boolean]> => {
+      const secureImageSource = src.replace(/^http:/, 'https:')
+      if (!(await urlIsBroken(secureImageSource))) {
+        shouldUpdate = true
+        return [src, secureImageSource]
+      }
+      return [src, false]
+    })
+  )).filter(tuple => tuple[1])) as Map<string, string>
```

Comment:
Maybe give tuple a more descriptive name, or add a docstring to say what this is doing. Took me a couple of reads to get the logic of this block.